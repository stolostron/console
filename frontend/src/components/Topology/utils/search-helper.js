/** *****************************************************************************
 * Licensed Materials - Property of IBM
 * (c) Copyright IBM Corporation 2018, 2019. All Rights Reserved.
 *
 * US Government Users Restricted Rights - Use, duplication or disclosure
 * restricted by GSA ADP Schedule Contract with IBM Corp.
 *******************************************************************************/
// Copyright (c) 2020 Red Hat, Inc.
// Copyright Contributors to the Open Cluster Management project

export const convertStringToQuery = searchText => {
  let relatedKinds = []

  if (searchText.indexOf('kind:subscription') >= 0) {
    relatedKinds = [
      'placementrule',
      'deployable',
      'application',
      'subscription',
      'channel'
    ]
  } else if (searchText.indexOf('kind:channel') >= 0) {
    relatedKinds = ['subscription']
  } else if (searchText.indexOf('kind:placementrule') >= 0) {
    relatedKinds = ['subscription']
  }

  const searchTokens = searchText.split(' ')
  const keywords = searchTokens.filter(
    token => token !== '' && token.indexOf(':') < 0
  )
  const filters = searchTokens
    .filter(token => token.indexOf(':') >= 0)
    .map(f => {
      const [property, values] = f.split(':')
      return { property, values: values.split(',') }
    })
    .filter(
      f =>
        ['', '=', '<', '>', '<=', '>=', '!=', '!'].findIndex(
          op => op === f.values[0]
        ) === -1
    )
  return { keywords, filters, relatedKinds }
}

export const formatNumber = count => {
  if (count > 999) {
    // show one decimal place
    const num = (count - count % 100) / 1000
    return `${num}k`
  }
  return count
}

const MAX_SEARCH_ATTEMPTS = 3
let searchFailureCount = 0
let searchErrorCount = 0

export const searchFailure = () => {
  searchFailureCount++
  searchErrorCount = 0
}
export const searchError = () => {
  searchErrorCount++
}
export const searchSuccess = () => {
  searchFailureCount = 0
  searchErrorCount = 0
}

export const shouldTrySearch = () =>
  searchFailureCount + searchErrorCount < MAX_SEARCH_ATTEMPTS

export const isSearchAvailable = () =>
  searchFailureCount === 0 && searchErrorCount === 0
export const isYAMLEditAvailable = () => searchErrorCount === 0
