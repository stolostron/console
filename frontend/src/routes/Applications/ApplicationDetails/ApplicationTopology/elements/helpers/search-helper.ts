/* Copyright Contributors to the Open Cluster Management project */

import type { SearchFilter, SearchQuery } from '../../types'

const MAX_SEARCH_ATTEMPTS = 3
let searchFailureCount = 0
let searchErrorCount = 0

/**
 * Parse a search bar string into keywords, filters, and inferred related kinds.
 * - Keywords are whitespace-separated tokens without a ':' character.
 * - Filters are tokens of the form "property:val1,val2".
 * - Operator-like values (e.g., '=', '>=') as the first value are filtered out.
 */
export const convertStringToQuery = (searchText: string): SearchQuery => {
  let relatedKinds: string[] = []

  if (searchText.indexOf('kind:subscription') >= 0) {
    relatedKinds = ['placementrule', 'deployable', 'application', 'subscription', 'channel']
  } else if (searchText.indexOf('kind:channel') >= 0) {
    relatedKinds = ['subscription']
  } else if (searchText.indexOf('kind:placementrule') >= 0) {
    relatedKinds = ['subscription']
  }

  const searchTokens = searchText.split(' ')
  const keywords = searchTokens.filter((token) => token !== '' && token.indexOf(':') < 0)
  const filters: SearchFilter[] = searchTokens
    .filter((token) => token.indexOf(':') >= 0)
    .map((f) => {
      const [property, values] = f.split(':')
      return { property, values: values.split(',') }
    })
    .filter((f) => ['', '=', '<', '>', '<=', '>=', '!=', '!'].findIndex((op) => op === f.values[0]) === -1)

  return { keywords, filters, relatedKinds }
}

/**
 * Format large numbers for display (e.g., 1000 => '1k').
 * Returns the original number if under 1000.
 */
export const formatNumber = (count: number): number | string => {
  if (count > 999) {
    // show one decimal place
    const num = (count - (count % 100)) / 1000
    return `${num}k`
  }
  return count
}

/** Track a search failure, resetting error streak. */
export const searchFailure = (): void => {
  searchFailureCount++
  searchErrorCount = 0
}

/** Track a search transport/error failure. */
export const searchError = (): void => {
  searchErrorCount++
}

/** Reset search failure/error counters on success. */
export const searchSuccess = (): void => {
  searchFailureCount = 0
  searchErrorCount = 0
}

/** Whether another search attempt should be made based on recent failures. */
export const shouldTrySearch = (): boolean => searchFailureCount + searchErrorCount < MAX_SEARCH_ATTEMPTS

/** Search UI is considered available if there have been no recent failures or errors. */
export const isSearchAvailable = (): boolean => searchFailureCount === 0 && searchErrorCount === 0

/** YAML edit is gated only by transport/errors, not logical failures. */
export const isYAMLEditAvailable = (): boolean => searchErrorCount === 0
