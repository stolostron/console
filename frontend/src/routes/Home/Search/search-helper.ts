/* Copyright Contributors to the Open Cluster Management project */
// Copyright (c) 2021 Red Hat, Inc.
// Copyright Contributors to the Open Cluster Management project

import { DropdownSuggestionsProps } from './components/Searchbar'

const operators = ['=', '<', '>', '<=', '>=', '!=', '!']
const dateValues = ['hour', 'day', 'week', 'month', 'year']

export function formatSearchbarSuggestions(
    data: string[],
    suggestionKind: 'label' | 'filter' | 'value',
    searchQuery: string
) {
    let valuesToRemoveFromSuggestions: string[] = []
    let suggestions: DropdownSuggestionsProps[] = []
    const labelTag = {
        id: 'id-suggestions-label',
        key: 'key-suggestions-label',
        name: 'Filters',
        kind: suggestionKind,
        disabled: true,
    }
    if (suggestionKind === 'value') {
        // Get a list of duplicate values to remove from suggestions dropdown
        const searchTokens = searchQuery.split(' ')
        const searchCompleteFilter = searchTokens[searchTokens.length - 1].replace(':', '')
        labelTag.name = `${searchCompleteFilter} values`
        const query = convertStringToQuery(searchQuery)
        query.filters.forEach((filter) => {
            if (filter.property === searchCompleteFilter) {
                valuesToRemoveFromSuggestions = filter.values.filter((value) => data.indexOf(value) > 0)
            }
        })
        if (data[0] === 'isNumber') {
            if (operators.some((operator: string) => searchQuery.endsWith(operator))) {
                // operator is already chosen
                const numberRange =
                    data.length > 2
                        ? `Min: ${parseInt(data[1], 10)} - Max: ${parseInt(data[2], 10)}`
                        : `Min: ${parseInt(data[1], 10)} - Max: ${parseInt(data[1], 10)}`

                return [
                    labelTag,
                    {
                        id: 'id-values-range',
                        key: 'key-values-range',
                        name: numberRange,
                        kind: suggestionKind,
                        disabled: true,
                    },
                ]
            }
            suggestions = operators.map((op) => {
                return {
                    id: `id-${op}`,
                    key: `key-${op}`,
                    name: op,
                    kind: suggestionKind,
                }
            })
            suggestions.unshift({
                id: 'id-operator-label',
                name: 'Operators',
                kind: 'label',
                disabled: true,
            })
            return suggestions
        } else if (data[0] === 'isDate') {
            suggestions = dateValues.map((date) => {
                return {
                    id: `id-date-${date}`,
                    key: `key-date-${date}`,
                    name: date,
                    kind: suggestionKind,
                }
            })
            suggestions.unshift({
                id: 'id-filter-label',
                name: `${searchCompleteFilter} within the last:`,
                kind: 'label',
                disabled: true,
            })
            return suggestions
        }
    }

    suggestions = data
        .filter((suggestion) => {
            return valuesToRemoveFromSuggestions.indexOf(suggestion) === -1
        })
        .map((field) => {
            return {
                id: `id-${field}`,
                key: `key-${field}`,
                name: field,
                kind: suggestionKind,
            }
        })
    suggestions.unshift(labelTag)
    return suggestions
}

export const convertStringToQuery = (searchText: string) => {
    const searchTokens = searchText.split(' ')
    const keywords = searchTokens.filter((token) => token !== '' && token.indexOf(':') < 0)
    const filters = searchTokens
        .filter((token) => token.indexOf(':') >= 0)
        .map((f) => {
            const splitIdx = f.indexOf(':')
            const property = f.substring(0, splitIdx)
            const values = f.substring(splitIdx + 1)
            return { property, values: values.split(',') }
        })
        .filter((f) => ['', '=', '<', '>', '<=', '>=', '!=', '!'].findIndex((op) => op === f.values[0]) === -1)
    return {
        keywords,
        filters,
        // Limit results to  for better performance - we show a info alert to user asking them to refine search.
        limit: 1000,
    }
}

export const getSearchCompleteString = (searchQuery: string) => {
    const queryTags = searchQuery.split(' ')
    if (queryTags[queryTags.length - 1].endsWith(':')) {
        return queryTags[queryTags.length - 1].replace(':', '')
    } else if (
        operators.some(
            (op) => queryTags[queryTags.length - 1].substring(queryTags[queryTags.length - 1].length - op.length) === op
        )
    ) {
        const operator = operators.filter(
            (op) => queryTags[queryTags.length - 1].substring(queryTags[queryTags.length - 1].length - op.length) === op
        )
        return queryTags[queryTags.length - 1].replace(':', '').replace(operator[operator.length - 1], '')
    }
    return ''
}
