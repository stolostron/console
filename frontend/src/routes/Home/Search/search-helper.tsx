/* Copyright Contributors to the Open Cluster Management project */
// Copyright (c) 2021 Red Hat, Inc.
// Copyright Contributors to the Open Cluster Management project

import { TFunction } from 'react-i18next'
import { Link } from 'react-router-dom-v5-compat'
import { NavigationPath } from '../../../NavigationPath'
import { SearchAlertProps } from './components/SearchAlertGroup'
import { DropdownSuggestionsProps } from './components/Searchbar'

export const federatedErrorText = 'error sending federated request'
export const operators = ['<=', '>=', '!=', '!', '=', '<', '>']
const dateValues = ['hour', 'day', 'week', 'month', 'year']

export function formatSearchbarSuggestions(
  data: string[],
  suggestionKind: 'label' | 'filter' | 'value',
  searchQuery: string,
  limit: number,
  isLoading: boolean,
  t: TFunction
) {
  const valuesToRemoveFromSuggestions: string[] = []
  let suggestions: DropdownSuggestionsProps[] = []
  const labelTag = {
    id: 'id-suggestions-label',
    name: t('Filters'),
    kind: suggestionKind,
    disabled: true,
  }
  if (suggestionKind === 'value') {
    // Get a list of duplicate values to remove from suggestions dropdown
    const searchTokens = searchQuery.split(' ')
    const searchCompleteFilter = searchTokens[searchTokens.length - 1].substring(
      0,
      searchTokens[searchTokens.length - 1].indexOf(':')
    )
    labelTag.name = t('{{0}} values', [searchCompleteFilter])
    const query = convertStringToQuery(searchQuery, limit)
    query.filters.forEach((filter) => {
      valuesToRemoveFromSuggestions.push(...(filter.property === searchCompleteFilter ? filter.values : []))
    })
    if (data[0] === 'isNumber') {
      if (operators.some((operator: string) => searchQuery.endsWith(operator))) {
        // operator is already chosen
        const numberRange =
          data.length > 2
            ? `Min: ${parseInt(data[1], 10)} - Max: ${parseInt(data[2], 10)}`
            : `Min: ${parseInt(data[1], 10)} - Max: ${parseInt(data[1], 10)}`

        return [
          labelTag,
          {
            id: 'id-values-range',
            name: numberRange,
            kind: suggestionKind,
            disabled: true,
          },
        ]
      }
      suggestions = operators.map((op) => {
        return {
          id: `id-${op}`,
          name: op,
          kind: suggestionKind,
        }
      })
      suggestions.unshift({
        id: 'id-operator-label',
        name: t('Operators'),
        kind: 'label',
        disabled: true,
      })
      return suggestions
    } else if (data[0] === 'isDate') {
      suggestions = dateValues.map((date) => {
        return {
          id: `id-date-${date}`,
          name: date,
          kind: suggestionKind,
        }
      })
      suggestions.unshift({
        id: 'id-filter-label',
        name: t('{{0}} within the last:', [searchCompleteFilter]),
        kind: 'label',
        disabled: true,
      })
      return suggestions
    }
  }

  suggestions = data
    .filter((suggestion) => {
      return valuesToRemoveFromSuggestions.indexOf(suggestion) === -1
    })
    .map((field) => {
      return {
        id: `id-${field}`,
        name: field,
        kind: suggestionKind,
      }
    })
  suggestions.unshift(labelTag)
  isLoading &&
    suggestions.push({
      id: 'id-loading',
      name: suggestionKind === 'filter' ? t('Loading...') : t('Loading more...'),
      kind: 'label',
      disabled: true,
    })
  return suggestions
}

export const convertStringToQuery = (searchText: string, queryResultLimit: number) => {
  const searchTokens = searchText.split(' ')
  const keywords = searchTokens.filter((token) => token !== '' && token.indexOf(':') < 0)
  const filters = searchTokens
    .filter((token) => token.indexOf(':') >= 0)
    .map((f) => {
      const splitIdx = f.indexOf(':')
      const property = f.substring(0, splitIdx)
      const values = f.substring(splitIdx + 1)
      return { property, values: values.split(',') }
    })

  return {
    keywords,
    filters,
    limit: queryResultLimit,
  }
}

export const getSearchCompleteString = (searchQuery: string) => {
  const queryTags = searchQuery.split(' ')
  if (queryTags[queryTags.length - 1].endsWith(':')) {
    return queryTags[queryTags.length - 1].replace(':', '')
  } else if (
    operators.some(
      (op) => queryTags[queryTags.length - 1].substring(queryTags[queryTags.length - 1].length - op.length) === op
    )
  ) {
    const operator = operators.filter(
      (op) => queryTags[queryTags.length - 1].substring(queryTags[queryTags.length - 1].length - op.length) === op
    )
    return queryTags[queryTags.length - 1].replace(':', '').replace(operator[0], '')
  }
  return ''
}

export function setFederatedErrorAlert(
  loading: boolean,
  error: any,
  data: any,
  alerts: SearchAlertProps[],
  addSearchAlert: (alert: SearchAlertProps) => void,
  removeSearchAlert: (key: string) => void,
  t: any
) {
  const federatedWarningKey = 'federated-error'
  if (!loading && error && error?.graphQLErrors.find((error: any) => error?.includes(federatedErrorText))) {
    !alerts.find((alert) => alert.key === federatedWarningKey) &&
      addSearchAlert({
        key: federatedWarningKey,
        variant: 'warning',
        title: t('One or more managed hubs are not reporting'),
        children: (
          <div style={{ display: 'flex', flexDirection: 'column' }}>
            {t(
              'Resources from these managed hubs and managed clusters will be missing in search results. Resolve cluster issues to see all cluster resources.'
            )}
            <Link style={{ paddingTop: '0.25rem' }} to={NavigationPath.clusters}>
              {t('View clusters')}
            </Link>
          </div>
        ),
      })
  } else if (
    !loading &&
    !error?.graphQLErrors.find((error: any) => error?.includes(federatedErrorText)) &&
    data &&
    alerts.find((alert) => alert.key === federatedWarningKey)
  ) {
    removeSearchAlert(federatedWarningKey)
  }
}
