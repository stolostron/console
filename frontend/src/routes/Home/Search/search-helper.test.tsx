/* Copyright Contributors to the Open Cluster Management project */
// Copyright (c) 2021 Red Hat, Inc.
// Copyright Contributors to the Open Cluster Management project
import { t } from 'i18next'
import { formatSearchbarSuggestions, convertStringToQuery, getSearchCompleteString } from './search-helper'

test('Correctly returns formatSearchbarSuggestions', () => {
  const testData = ['kind', 'cluster', 'deployment']
  const result = formatSearchbarSuggestions(testData, 'filter', '', t)
  expect(result).toMatchSnapshot()
})

test('Correctly returns formatSearchbarSuggestions of name values', () => {
  const testData = ['name1', 'name2', 'name3']
  const result = formatSearchbarSuggestions(testData, 'value', 'name:name1 name:', t)
  expect(result).toMatchSnapshot()
})

test('Correctly returns formatSearchbarSuggestions of namespace values', () => {
  const testData = ['namespace1', 'namespace2', 'namespace3']
  const result = formatSearchbarSuggestions(testData, 'value', 'name:name1 namespace:', t)
  expect(result).toMatchSnapshot()
})

test('Correctly returns formatSearchbarSuggestions of operators', () => {
  const testData = ['isNumber']
  const result = formatSearchbarSuggestions(testData, 'value', 'cpu:', t)
  expect(result).toMatchSnapshot()
})

test('Correctly returns formatSearchbarSuggestions with operator already chosen', () => {
  const testData = ['isNumber', '1', '10']
  const result = formatSearchbarSuggestions(testData, 'value', 'cpu:=', t)
  expect(result).toMatchSnapshot()
})

test('Correctly returns formatSearchbarSuggestions with operator already chosen (single number option)', () => {
  const testData = ['isNumber', '1']
  const result = formatSearchbarSuggestions(testData, 'value', 'cpu:=', t)
  expect(result).toMatchSnapshot()
})

test('Correctly returns formatSearchbarSuggestions with date', () => {
  const testData = ['isDate']
  const result = formatSearchbarSuggestions(testData, 'value', 'created:', t)
  expect(result).toMatchSnapshot()
})

test('Correctly returns convertStringToQuery', () => {
  const testData = 'namespace:open-cluster-management kind:pod'
  const result = convertStringToQuery(testData)
  expect(result).toMatchSnapshot()
})

test('Correctly returns getSearchCompleteString', () => {
  const testData = 'namespace:open-cluster-management kind:'
  const result = getSearchCompleteString(testData)
  expect(result).toMatchSnapshot()
})

test('Correctly returns getSearchCompleteString with operator', () => {
  const testData = 'cpu:='
  const result = getSearchCompleteString(testData)
  expect(result).toMatchSnapshot()
})
