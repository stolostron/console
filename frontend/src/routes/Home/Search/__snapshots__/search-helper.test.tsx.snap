// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Correctly returns convertStringToQuery 1`] = `
Object {
  "filters": Array [
    Object {
      "property": "namespace",
      "values": Array [
        "open-cluster-management",
      ],
    },
    Object {
      "property": "kind",
      "values": Array [
        "pod",
      ],
    },
  ],
  "keywords": Array [],
  "limit": 1000,
}
`;

exports[`Correctly returns formatSearchbarSuggestions 1`] = `
Array [
  Object {
    "disabled": true,
    "id": "id-suggestions-label",
    "kind": "filter",
    "name": "Filters",
  },
  Object {
    "id": "id-kind",
    "kind": "filter",
    "name": "kind",
  },
  Object {
    "id": "id-cluster",
    "kind": "filter",
    "name": "cluster",
  },
  Object {
    "id": "id-deployment",
    "kind": "filter",
    "name": "deployment",
  },
]
`;

exports[`Correctly returns formatSearchbarSuggestions of name values 1`] = `
Array [
  Object {
    "disabled": true,
    "id": "id-suggestions-label",
    "kind": "value",
    "name": "name values",
  },
  Object {
    "id": "id-name1",
    "kind": "value",
    "name": "name1",
  },
  Object {
    "id": "id-name2",
    "kind": "value",
    "name": "name2",
  },
  Object {
    "id": "id-name3",
    "kind": "value",
    "name": "name3",
  },
]
`;

exports[`Correctly returns formatSearchbarSuggestions of namespace values 1`] = `
Array [
  Object {
    "disabled": true,
    "id": "id-suggestions-label",
    "kind": "value",
    "name": "namespace values",
  },
  Object {
    "id": "id-namespace1",
    "kind": "value",
    "name": "namespace1",
  },
  Object {
    "id": "id-namespace2",
    "kind": "value",
    "name": "namespace2",
  },
  Object {
    "id": "id-namespace3",
    "kind": "value",
    "name": "namespace3",
  },
]
`;

exports[`Correctly returns formatSearchbarSuggestions of operators 1`] = `
Array [
  Object {
    "disabled": true,
    "id": "id-operator-label",
    "kind": "label",
    "name": "Operators",
  },
  Object {
    "id": "id-<=",
    "kind": "value",
    "name": "<=",
  },
  Object {
    "id": "id->=",
    "kind": "value",
    "name": ">=",
  },
  Object {
    "id": "id-!=",
    "kind": "value",
    "name": "!=",
  },
  Object {
    "id": "id-!",
    "kind": "value",
    "name": "!",
  },
  Object {
    "id": "id-=",
    "kind": "value",
    "name": "=",
  },
  Object {
    "id": "id-<",
    "kind": "value",
    "name": "<",
  },
  Object {
    "id": "id->",
    "kind": "value",
    "name": ">",
  },
]
`;

exports[`Correctly returns formatSearchbarSuggestions with date 1`] = `
Array [
  Object {
    "disabled": true,
    "id": "id-filter-label",
    "kind": "label",
    "name": "created within the last:",
  },
  Object {
    "id": "id-date-hour",
    "kind": "value",
    "name": "hour",
  },
  Object {
    "id": "id-date-day",
    "kind": "value",
    "name": "day",
  },
  Object {
    "id": "id-date-week",
    "kind": "value",
    "name": "week",
  },
  Object {
    "id": "id-date-month",
    "kind": "value",
    "name": "month",
  },
  Object {
    "id": "id-date-year",
    "kind": "value",
    "name": "year",
  },
]
`;

exports[`Correctly returns formatSearchbarSuggestions with operator already chosen (single number option) 1`] = `
Array [
  Object {
    "disabled": true,
    "id": "id-suggestions-label",
    "kind": "value",
    "name": "cpu values",
  },
  Object {
    "disabled": true,
    "id": "id-values-range",
    "kind": "value",
    "name": "Min: 1 - Max: 1",
  },
]
`;

exports[`Correctly returns formatSearchbarSuggestions with operator already chosen 1`] = `
Array [
  Object {
    "disabled": true,
    "id": "id-suggestions-label",
    "kind": "value",
    "name": "cpu values",
  },
  Object {
    "disabled": true,
    "id": "id-values-range",
    "kind": "value",
    "name": "Min: 1 - Max: 10",
  },
]
`;

exports[`Correctly returns getSearchCompleteString 1`] = `"kind"`;

exports[`Correctly returns getSearchCompleteString with operator 1`] = `"cpu"`;
